components/ai_chat/core/browser/model_service.cc

```diff
@@ -0,0 +1,403 @@
+
+  model->key = base::StrCat(
+      {"custom:",
+       base::Uuid::GenerateRandomV4().AsLowercaseString().substr(0, 8)});
+
+  base::Value::List custom_models_pref =
+      pref_service_->GetList(prefs::kCustomModelsList).Clone();
+  base::Value::Dict model_dict = GetModelDict(std::move(model));
+  custom_models_pref.Append(std::move(model_dict));
+  pref_service_->SetList(prefs::kCustomModelsList,
+                         std::move(custom_models_pref));
+
+  InitModels();
+}
+
+void ModelService::SaveCustomModel(uint32_t index, mojom::ModelPtr model) {
+  base::Value::List custom_models_pref =
+      pref_service_->GetList(prefs::kCustomModelsList).Clone();
+
+  if (index >= custom_models_pref.size() || index < 0) {
+    return;
+  }
+
+  auto model_iter = custom_models_pref.begin() + index;
+
+  const std::string& existing_key =
+      *model_iter->GetDict().FindString(kCustomModelItemKey);
+
+  // Make sure the key is not changed when modifying the model
+  // because Dict::Merge is destructive.
+  CHECK(existing_key == model->key)
+      << "Model key mismatch. Existing key: " << existing_key
+      << ", sent model key: " << model->key << ".";
+
+  base::Value::Dict model_dict = GetModelDict(std::move(model));
+  model_iter->GetDict().Merge(std::move(model_dict));
+
+  pref_service_->SetList(prefs::kCustomModelsList,
+                         std::move(custom_models_pref));
+
+  InitModels();
+}
+
+void ModelService::DeleteCustomModel(uint32_t index) {
+  base::Value::List custom_models_pref =
+      pref_service_->GetList(prefs::kCustomModelsList).Clone();
+
+  if (index >= custom_models_pref.size() || index < 0) {
+    return;
+  }
+
+  auto model = custom_models_pref.begin() + index;
+  std::string removed_key = *model->GetDict().FindString(kCustomModelItemKey);
+
+  custom_models_pref.erase(model);
+  pref_service_->SetList(prefs::kCustomModelsList,
+                         std::move(custom_models_pref));
+
+  auto current_default_key = pref_service_->GetString(prefs::kDefaultModelKey);
+
+  // If the removed model is the default model, clear the default model key.
+  if (current_default_key == removed_key) {
+    pref_service_->ClearPref(prefs::kDefaultModelKey);
+    DVLOG(1) << "Default model key " << removed_key
+             << " was removed. Cleared default model key.";
+  }
+
+  InitModels();
+  for (auto& obs : observers_) {
+    obs.OnModelRemoved(removed_key);
+    obs.OnDefaultModelChanged(features::kAIModelsDefaultKey.Get());
+  }
+}
+
+void ModelService::ChangeDefaultModel(const std::string& new_key) {
+  const auto& models = GetModels();
+
+  bool does_model_exist = base::Contains(
+      models, new_key, [](const mojom::ModelPtr& model) { return model->key; });
+
+  if (!does_model_exist) {
+    DVLOG(1) << "Default model key " << new_key
+             << " does not exist in the model list.";
+    return;
+  }
+
+  pref_service_->SetString(prefs::kDefaultModelKey, new_key);
```

Comment:
Ok, this makes sense but how does on resolve access to the key in files like `brave/browser/extensions/api/settings_private/brave_prefs_util.cc` ?

Reply:
alright, looks like we don't need an entry in `brave_prefs_util.cc` anymore because all mutations to `kDefaultModelKey` go through the ModelService now.